packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts

```diff
@@ -121,6 +138,123 @@ export async function buildPgDatabaseConfig(
   };
 }
 
+/* Note: the following type definition is intentionally duplicated in
+ * /packages/backend-defaults/config.d.ts so the clientSecret property
+ * can be annotated with "@visibility secret" there.
+ */
+export type AzureTokenCredentialConfig = {
+  /**
+   * How early before an access token expires to refresh it with a new one.
+   * Defaults to 5 minutes
+   * Supported formats:
+   * - A string in the format of '1d', '2 seconds' etc. as supported by the `ms`
+   *   library.
+   * - A standard ISO formatted duration string, e.g. 'P2DT6H' or 'PT1M'.
+   * - An object with individual units (in plural) as keys, e.g. `{ days: 2, hours: 6 }`.
+   */
+  tokenRenewalOffsetTime?: string | HumanDuration;
+} & (
+  | {
+      type: 'ClientSecretCredential';
+      clientId: string;
+      clientSecret: string;
+      tenantId: string;
+    }
+  | {
+      type: 'ManagedIdentityCredential';
+      /**
+       * The client ID of a user-assigned managed identity.
+       * If not provided, the system-assigned managed identity is used.
+       */
+      clientId?: string;
+    }
+  | {
+      type?: undefined | 'DefaultAzureCredential';
+    }
+);
+
+export async function buildAzurePgConfig(config: any, sanitizedConfig: any) {
+  const {
+    DefaultAzureCredential,
+    ManagedIdentityCredential,
+    ClientSecretCredential,
+  } = require('@azure/identity');
+
+  // By default get a new token starting five minutes before the old one expires
+  let tokenRenewalOffsetMs = 300_000;
+  const configReader = new ConfigReader(config);
+  if (configReader.has('connection.tokenCredential.tokenRenewalOffsetTime')) {
+    tokenRenewalOffsetMs = durationToMilliseconds(
+      readDurationFromConfig(configReader, {
+        key: 'connection.tokenCredential.tokenRenewalOffsetTime',
+      }),
+    );
+  }
+
+  let credential: TokenCredential;
+  if (configReader.has('connection.tokenCredential')) {
+    const tokenCredentialConfig = configReader.get(
+      'connection.tokenCredential',
+    ) as AzureTokenCredentialConfig;
+    switch (tokenCredentialConfig.type) {
```

Comment:
Does this mean the `type` field has to be specified in the configuration file? In the Azure integration the this `type` inferred based on the fields you configure. For example if you pass `tenant_id` + `client_id` + `client_secret` we assume that it is a `ClientSecretCredential`, if you pass `tenant_id` + `clien_id` we assume it is a `ManagedIdentityCredential`

Reply:
My personal preference is to require the type to be explicit, but I can see that it could be inferred at least for now until something changes and it becomes ambiguous.